[CF-490] Support for cinder migration with re-attach of backend to destination Cloud

SVilgelm opened the pull request:
[CF-496] Use juno nfs folders for all boxes
[CF-498] Migrate volume with reattaching old files
[CF-490] Make a dump of databases
[CF-490] Mute stdout of fabric
[CF-490] Use override rules to override volume_type
[CF-490] Delete failed volume

SVilgelm committed:
[CF-496] Use juno nfs share

Use juno nfs folders for all boxes

MirantisWorkloadMobility commented:
There were no python files updated. Admins, please, check pull request.

roman-verchikov commented on cloudferry_devlab/provision/configure_nfs.sh:
```diff
@@ -0,0 +1,6 @@
+
+CINDER_CONF_FILES="/etc/cinder/nfsshares*"
+
+sed -i "s/${1}/juno/g" ${CINDER_CONF_FILES}
```
`${1}` is unreadable, please assign to a local variable

SVilgelm replied:
Done

roman-verchikov commented on cloudferry_devlab/Vagrantfile:
```diff
@@ -101,6 +101,12 @@ Vagrant.configure(2) do |config|
           end
           thisnode.vm.provision "shell", path: "./provision/libvirt.sh"
+
+          if ENV["SAME_NFS_SHARE"] and nodename != "juno" then
```
I'd suggest renaming to `SHARED_NFS_BACKEND`

SVilgelm replied:
Done

MirantisWorkloadMobility commented:
There were no python files updated. Admins, please, check pull request.

roman-verchikov commented:
+1

MirantisWorkloadMobility commented:
There were no python files updated. Admins, please, check pull request.

SVilgelm committed:
[CF-498] Migrate volume with reattaching old files

1. Create a fake volume on dst with a minimum possible size (1G) and
with the same parameters as source volume.
2. Remove the file of fake volume.
3. Update ID of the fake volume and its size directly in the database.

SVilgelm committed:
[CF-490] Make a dump of databases

MirantisWorkloadMobility commented:
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.storage.cinder_db
cloudferry/lib/os/storage/cinder_db.py:23: [W0621(redefined-outer-name), 
CinderVolume.__init__] Redefining name 'uuid' from outer scope (line 15)
************* Module cloudferry.cfglib
cloudferry/cfglib.py:692: [E0602(undefined-variable), ] Undefined variable 'snapshot'
cloudferry/cfglib.py:692: [E0602(undefined-variable), ] Undefined variable 
'snapshot_opts'

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3817/

roman-verchikov commented on cloudferry/actions/block_storage/migrate_volumes.py:
```diff
@@ -116,6 +122,34 @@ def __init__(self, init):
+            return True
+        elif src_shared_nfs or dst_shared_nfs:
+            msg = ("Invalid configuration of src or dst storage backends. "
+                   "In case OF shared-nfs both options of the backends must "
```
Change "OF" to lowercase

SVilgelm replied:
Done

roman-verchikov commented on cloudferry/actions/block_storage/migrate_volumes.py:
```diff
@@ -215,13 +249,46 @@ def run(self, **kwargs):
+                     volume_name(src_volume))
+            dst_cinder = self.dst_cloud.resources[utils.STORAGE_RESOURCE]
+
+            dst_volume = dst_cinder.get_migrated_volume(src_volume['id'])
```
You'd need to modify `get_migrated_volume` to look for volume with the same ID for 'shared-nfs' case, and keep existing implementation for other backends.

SVilgelm replied:
No, it works fine. It uses metadata to match the  volumes.

roman-verchikov commented on cloudferry/cfglib.py:
```diff
@@ -309,7 +309,7 @@
 src_storage_opts = [
     cfg.StrOpt('backend', default='nfs',
-               choices=['nfs', 'iscsi-vmax'],
+               choices=['nfs', 'iscsi-vmax', 'shared-nfs'],
```
Please update documentation too

SVilgelm replied:
Done

roman-verchikov commented on cloudferry/lib/utils/mysql_connector.py:
```diff
@@ -36,6 +41,29 @@ def get_db_host(cloud_config):
+                  user=cloud.cloud_config.mysql.db_user,
+                  password=cloud.cloud_config.mysql.db_password)
+
+    filename = "dump_{position}_{database}_{time}.sql"
```
I'd suggest this value to be configurable

SVilgelm replied:
Done

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3821/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

SVilgelm commented:
It works: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/26/

SVilgelm commented:
retest

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

roman-verchikov commented on cloudferry/lib/utils/mysql_connector.py:
```diff
@@ -36,6 +41,29 @@ def get_db_host(cloud_config):
 
 
+def dump_db(cloud, database=ALL_DATABASES):
+    cmd = "mysqldump {database} --user={user}"
```
This command produces tons of unnecessary logs, please consider calling `mysqldump` remotely

SVilgelm committed:
[CF-490] Mute stdout of fabric

Allow do not write messages from fabric's stdout to the logger.

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3844/

MirantisWorkloadMobility commented:
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.log
cloudferry/lib/utils/log.py:45: [W0212(protected-access), StdoutLogger.write] Access to 
a protected member _mute of a client class
cloudferry/lib/utils/log.py:23: [W0611(unused-import), ] Unused import threading

MirantisWorkloadMobility commented:
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.log
cloudferry/lib/utils/log.py:44: [W0212(protected-access), StdoutLogger.write] Access to 
a protected member _mute of a client class

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

SVilgelm commented:
http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/34/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

antonf commented on cloudferry/lib/utils/override.py:
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
-        return cls(data)
+            data = data.get(object_type, {})
+            if not isinstance(data, dict):
+                raise TypeError('%s root object for %s must be dictionary!' %
```
this is not root object

SVilgelm replied:
Done

antonf commented on cloudferry/lib/utils/override.py:
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
                 raise TypeError('%s root object must be dictionary!' %
                                 (mapping_filename,))
-        return cls(data)
+            data = data.get(object_type, {})
```
Please use different variable name

SVilgelm replied:
Done

antonf commented:
👎

roman-verchikov commented on cloudferry/lib/os/storage/cinder_storage.py:
```diff
@@ -341,6 +345,11 @@ def create_volume(self, size, **kwargs):
         cinder = self.cinder_client
         tenant_id = kwargs.get('project_id')
 
+        if self.cloud.position == 'dst':
```
I think `volume_type` should be treated equally for both source and destination, I'd suggest removing this check

SVilgelm replied:
OK. Done

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
$PULL_STATUS
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3859/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

SVilgelm commented:
retest

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
There were no python files updated. Admins, please, check pull request.

antonf commented on cloudferry/lib/utils/override.py:
```diff
@@ -70,18 +70,29 @@ def __init__(self, mapping):
-        return cls(data)
+            object_data = data.get(object_type, {})
+            if not isinstance(object_data, dict):
+                raise TypeError('%s object must be dictionary!' %
```
Please specify the file: '%s object in mapping file %s must be dictionary!'

SVilgelm replied:
Done

antonf commented:
👎

MirantisWorkloadMobility commented:
$PYLINT_STATUS

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

roman-verchikov commented:
please rebase

MirantisWorkloadMobility commented:
$PYLINT_STATUS

MirantisWorkloadMobility commented:
$PULL_STATUS
$BUILD_URL

roman-verchikov commented:
-1

1. We must not leave DB in undefined state in case when DB error occurs for any reason;
2. Rollback procedure for failed volume should be kept the same as previously: if failure occurs, created volume must be removed and warning should be printed.

SVilgelm committed:
[CF-490] Delete failed volume

Use transactions for mysql queries.

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

SVilgelm commented:
works with shared-nfs: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/39/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

archyufa commented:
Migration of single unattached volume worked in staging.

archyufa commented:
Can the backup of cinder databases be written to the file ? At the moment it is just dumps on screen.

SVilgelm commented:
I'm working on it now

archyufa commented:
Tested in staging. It looks like quota is not working correctly. As it only increase quota per 1 GB every time, even if volume is more than 1 GB.

MirantisWorkloadMobility commented:
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3941/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3943/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3948/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3949/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

SVilgelm commented:
retest

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

archyufa commented:
+1 tested in staging

archyufa commented:
can be merged

archyufa commented on configs/config.ini:
```diff
@@ -208,6 +208,12 @@ migrate_whole_cloud = False
+# Filename template to store a dump of a database.
+db_dump_filename = dump_{position}_{database}_{time}.sql
+
+# Run mysqldump command on database's host. Otherwise mysqldump will be
```
a bit not clear. If set to True it will run command on database's host?

SVilgelm replied:
yes, or on the `migration.mysqldump_host` if it is set.

archyufa commented:
It will be good to have possibility to disable snapshot. In case problem with it. Also there will be no need to take snapshots every time as this dumps consume space and takes extra time to run every time.

roman-verchikov commented on cloudferry/actions/block_storage/migrate_volumes.py:
```diff
@@ -205,24 +232,74 @@ def _create_volume(self, src_volume):
+                        retrying.TimeoutExceeded,
+                        exception.TenantNotPresentInDestination,
+                        cinder_exceptions.OverLimit,
+                        cinder_exceptions.NotFound,
```
Let's catch all cinder exceptions here

SVilgelm replied:
Done

roman-verchikov commented:
-1, please update exception handling

antonf commented:
> It will be good to have possibility to disable snapshot

Are you serious? I am surprised that person working with production environment request such a "feature". In case there is problem with making database snapshot you should fix this instead of disabling making backup of your database. Working with OpenStack database directly is extremely dangerous and you should have backup of production database as recent as possible *each time*. Production database can change at any time: even when you run your migration for tenant A, some user can create volume in tenant B, for example.

While it is possible to do what you ask, we shouldn't do that.

SVilgelm commented:
retest

SVilgelm commented:
It works: http://jenkins-cf.vm.mirantis.net:8080/view/Functional%20test/job/cloudferry-functional-tests-grizzly-to-juno-shared-nfs/53/

MirantisWorkloadMobility commented:
PEP8 check success, Unittests success

roman-verchikov commented:
can be merged
